Improve testing documentation in `CONTRIBUTING.md` (#54904)
pytorch/pytorch: CONTRIBUTING.md

- [Nightly Checkout & Pull](#nightly-checkout--pull)
- [Codebase structure](#codebase-structure)
- [Unit testing](#unit-testing)

>     - [Python Unit Testing](#python-unit-testing)
>     - [Better local unit tests with pytest](#better-local-unit-tests-with-pytest)
>     - [Running `mypy`](#running-mypy)
>     - [C++ Unit Testing](#c-unit-testing)

Reviewer: Sorry for the inconvenience; this just needs a couple minor formatting tweaks to fix the failing ["Lint / toc" job](https://github.com/pytorch/pytorch/pull/54904/checks?check_run_id=2242141332):
```suggestion
  - [Python Unit Testing](#python-unit-testing)
  - [Better local unit tests with `pytest`](#better-local-unit-tests-with-pytest)
  - [Running `mypy`](#running-mypy)
  - [C++ Unit Testing](#c-unit-testing)
```

name of an individual test with `TESTCLASSNAME.TESTNAME`. Here,
`TESTNAME` is the name of the test you want to run, and `TESTCLASSNAME`
is the name of the class in which it is defined.

Going off the above example, let's say you want to run
`test_Sequential`, which is defined as part of the `TestJit` class
in `test/test_jit.py`. Your command would be:

```bash
python test/test_jit.py TestJit.test_Sequential
```

The `hypothesis` library must be installed to run the tests. `mypy` is
an optional dependency, and `pytest` may help run tests more selectively.
All these packages can be installed with `conda` or `pip`.

### Better local unit tests with `pytest`
We don't officially support `pytest`, but it works well with our
`unittest` tests and offers a number of useful features for local
developing. Install it via `pip install pytest`.

If you want to just run tests that contain a specific substring, you can
use the `-k` flag:

```bash
pytest test/test_nn.py -k Loss -v
```

The above is an example of testing a change to all Loss functions: this
command runs tests such as `TestNN.test_BCELoss` and
`TestNN.test_MSELoss` and can be useful to save keystrokes.

### Running `mypy`

> `mypy` is an optional static type checker for Python. We have a test
> suite that runs `mypy` on the entire codebase:
> 
> ```bash
> python test/test_type_hints.py
> ```
> 

Reviewer: This is actually out of date; while we're updating this section, could we update this as well?
````suggestion
`mypy` is an optional static type checker for Python. We have multiple `mypy`
configs for the PyTorch codebase, so you can run them all using this command:

```bash
for CONFIG in mypy*.ini; do mypy --config="$CONFIG"; done
```

````

See [Guide for adding type annotations to
PyTorch](https://github.com/pytorch/pytorch/wiki/Guide-for-adding-type-annotations-to-PyTorch)

> for more information on how to set up `mypy` and tackle type annotation
> tasks, as well as other ways to run `mypy` besides running that test suite.

Reviewer: Similarly here:
```suggestion
for more information on how to set up `mypy` and tackle type annotation tasks.
```
